Improvements for JavaScript bindings

ricardoquesada opened the pull request:
Moves rooting to one function so it will be easier to migrate to the GC model.

ricardoquesada committed:
my JS changes... again

ricardoquesada committed:
fixes autorelease issues on rebind

ricardoquesada committed:
cherry-picking changes to disable retain/release from GC

ricardoquesada committed:
RepeatForever, Repeat and Speed converted

ricardoquesada commented:
It is failing because I haven't updated the submodules, but if the PR is good for you ( @pandamicro  ), I'll update the submodules.

pandamicro commented on cocos/scripting/js-bindings/manual/ScriptingCore.cpp:
```diff
@@ -1960,4 +1985,67 @@ void jsb_remove_proxy(js_proxy_t* nativeProxy, js_proxy_t* jsProxy)
+#else
+    ref->autorelease();
+    JS::AddNamedObjectRoot(cx, obj, debug);
+#endif
```
Is the macro control inverse ? I think we only should add root for JS object when native control script, not reversely.

ricardoquesada replied:
ha... well, I know what you mean.
"native" for me means: the native way for JS... but I'll rename it to avoid confusion.

pandamicro replied:
:joy: Ok, I see

pandamicro commented on cocos/scripting/js-bindings/manual/ScriptingCore.cpp:
```diff
@@ -1960,4 +1985,67 @@ void jsb_remove_proxy(js_proxy_t* nativeProxy, js_proxy_t* jsProxy)
+            CCLOG("jsb_ref_finalize: JSObject address = %p (%s)", obj, typeid(*ref).name());
 
+            jsb_remove_proxy(nproxy, jsproxy);
+            ref->release();
```
If native control script, I don't think we should release the native object, am I missing something here ?

ricardoquesada replied:
native control is disabled by default. The code is there so that we could keep doing our tests.

ricardoquesada committed:
better macro name for GC. Thanks panda

ricardoquesada commented:
@pandamicro  thanks. I fixed the macro name. New name is: `CC_ENABLE_GC_FOR_NATIVE_OBJECTS`

pandamicro commented:
Thanks, I will try to understand the essential idea of your PR and give my feed backs

ricardoquesada commented:
Ok. thanks. 
My PR is pretty simple... it just remove some Root/Unroot from the bindings, and moved it to `jsb_ref_init()` and `jsb_ref_rebind()`...and also converted some `Node::create`, to `new Node;node->init()`

pandamicro commented on cocos/base/CCRef.cpp:
```diff
@@ -83,6 +91,18 @@ void Ref::retain()
+        auto scriptMgr = ScriptEngineManager::getInstance()->getScriptEngine();
+        if (scriptMgr && scriptMgr->getScriptType() == kScriptTypeJavascript)
+        {
+            scriptMgr->rootRef(this);
```
I think you mean rootObject and unrootObject below

ricardoquesada replied:
nice catch. thanks.

ricardoquesada committed:
ooops... thanks Panda

pandamicro commented:
I think we should merge: 
- js_get_type_from_native
- jsb_ref_create_jsobject
- jsb_ref_init
- jsb_ref_rebind
- jsb_ref_finalize

It can greatly simplify bindings code ! It should be shipped with v3.10.

As for `CC_ENABLE_GC_FOR_NATIVE_OBJECTS`, can you remove them from this PR ? We should do more test and try to have a stable solution for GC memory model, then we can create another PR for it.

What do you think ? @ricardoquesada

ricardoquesada committed:
updated submodule

ricardoquesada commented:
My opinion is:
- we shouldn't remove `CC_ENABLE_GC_FOR_NATIVE_OBJECTS` from the PR
- we should ship for v3.10
- it should be disabled by default.
- QA should not test it
- Only you, me, another member of the community who are interested in this feature should test it.
- we should mention on the v3.10 press release that we have an experimental GC solution for JS (and Lua?)

ricardoquesada commented:
Squashed everything in one commit here:
https://github.com/cocos2d/cocos2d-x/pull/14553

ricardoquesada commented:
@pandamicro I merged the PR (squashed) with `CC_ENABLE_GC_FOR_NATIVE_OBJECTS` but I can remove it if you want.

I think the next steps are:
- Make sure there are not manual calls to `AddRoot` /  `RemoveRoot`... instead we should replace then with `jsb_ref_init` / `jsb_ref_finalize`
- Make sure there not `XXX::create()` calls when creating new objects... instead we should replace them with `new` and `init`

Once we have that, I think we will be ready to test GC for native objects

pandamicro commented:
Ok, I will check for AddRoot / RemoveRoot usage after submit v3.10 to QA.
